src/frontend/state/AchievementState.ts

```diff
@@ -0,0 +1,312 @@
+    }
+    const keyString = JSON.stringify(key)
+    return keyString
+  }
+
+  async fetchIndividualAchievements({
+    gameId,
+    page,
+    pageSize
+  }: {
+    gameId: string
+    page: number
+    pageSize: number
+  }) {
+    const keyString = this.individualAchievementsKey(gameId)
+
+    /**
+     * Check cache for hit
+     */
+    if (this.individualAchievements.has(keyString)) {
+      return this.individualAchievements.get(keyString)?.data
+    }
+
+    const individualAchievements = await window.api.getIndividualAchievements({
+      gameId: Number(gameId),
+      store: this.store,
+      sort: this.currentIndividualSort.value,
+      page,
+      pageSize,
+      playerStoreId: this.playerStoreId,
+      playerAddress: walletState.address
+    })
+
+    this.individualAchievements.set(keyString, individualAchievements)
+    return individualAchievements.data
+  }
+
+  individualAchievementsForGame(gameId: string) {
+    const keyString = this.individualAchievementsKey(gameId)
+    return this.individualAchievements.get(keyString)
+  }
+
+  fetchMoreSummaryAchievements = async () => {
+    if (this.fetching) {
+      return
+    }
+    if (!this.playerStoreId) {
+      console.error('player store id is not set!')
+      return
+    }
+    this.setFetching(true)
+    const pagesFetched = this.currentSummaryAchievements?.pagesFetched
+    const totalPages = this.currentSummaryAchievements?.totalPages
+    const alreadyFetchedAllPages =
+      pagesFetched !== undefined &&
+      totalPages !== undefined &&
+      pagesFetched >= totalPages
+    if (alreadyFetchedAllPages) {
+      this.setFetching(false)
+      return
+    }
+    const requestKey = this.currentSummaryKey
+
+    const nextPageOfSummaryAchievements =
+      await window.api.getSummaryAchievements({
+        store: this.store,
+        filter: requestKey.filter,
+        sort: requestKey.sort,
+        // api is one indexed not zero indexed
+        page: pagesFetched ? pagesFetched + 1 : 1,
+        playerStoreId: this.playerStoreId,
+        playerAddress: walletState.address
+      })
+
+    const valRef = this.summaryAchievements.get(JSON.stringify(requestKey))
+
+    if (valRef !== undefined) {
+      valRef.data = valRef.data.concat(nextPageOfSummaryAchievements.data)
+      valRef.pagesFetched += 1
+    } else {
+      this.summaryAchievements.set(JSON.stringify(requestKey), {
+        data: nextPageOfSummaryAchievements.data,
+        pagesFetched: 1,
+        totalPages: nextPageOfSummaryAchievements.totalPages
+      })
+    }
+    this.setFetching(false)
+  }
+
+  /**
+   * Necessary since strict-mode is enabled in mobx
+   */
+  setFetching = (val: boolean) => {
+    runInAction(() => {
+      this.fetching = val
+    })
+  }
+
+  private get currentSummaryKey() {
+    const requestKey: AchievementRequestKey = {
+      filter: this.currentFilter,
+      sort: this.currentSort.value,
+      store: this.currentStore
+    }
+    return requestKey
```

Comment:
Could be smaller method but optional

```typescript
private get currentSummaryKey(): AchievementRequestKey  {
    return {
      filter: this.currentFilter,
      sort: this.currentSort.value,
      store: this.currentStore
    }
}
````

Reply:
okay updated